src/vm/types/mod.rs

```diff
@@ -0,0 +1,450 @@
+            }
+        }
+    }
+}
+
+impl From<StandardPrincipalData> for Value {
+    fn from(principal: StandardPrincipalData) -> Self {
+        Value::Principal(PrincipalData::from(principal))
+    }
+}
+
+impl From<PrincipalData> for Value {
+    fn from(p: PrincipalData) -> Self {
+        Value::Principal(p)
+    }
+}
+
+impl From<StandardPrincipalData> for PrincipalData {
+    fn from(p: StandardPrincipalData) -> Self {
+        PrincipalData::StandardPrincipal(p)
+    }
+}
+
+impl From<TupleData> for Value {
+    fn from(t: TupleData) -> Self {
+        Value::Tuple(t)
+    }
+}
+
+impl TupleData {
+    fn new(type_signature: TupleTypeSignature, data_map: BTreeMap<ClarityName, Value>) -> Result<TupleData> {
+        let t = TupleData { type_signature, data_map };
+        if t.size()? > MAX_VALUE_SIZE {
+            return Err(RuntimeErrorType::ValueTooLarge.into())
+        }
+        Ok(t)
+    }
+    pub fn from_data(mut data: Vec<(ClarityName, Value)>) -> Result<TupleData> {
+        let mut type_map = BTreeMap::new();
+        let mut data_map = BTreeMap::new();
+        for (name, value) in data.drain(..) {
+            let type_info = TypeSignature::type_of(&value);
+            if type_map.contains_key(&name) {
+                return Err(UncheckedError::VariableDefinedMultipleTimes(name.into()).into());
+            } else {
+                type_map.insert(name.clone(), type_info);
+            }
+            data_map.insert(name, value);
+        }
+
+        Self::new(TupleTypeSignature { type_map }, data_map)
+    }
+
+    pub fn from_data_typed(mut data: Vec<(ClarityName, Value)>, expected: &TupleTypeSignature) -> Result<TupleData> {
+        let type_map = &expected.type_map;
+        let mut data_map = BTreeMap::new();
+        for (name, value) in data.drain(..) {
+            let expected_type = type_map.get(&name)
+                .ok_or(InterpreterError::FailureConstructingTupleWithType)?;
+            if !expected_type.admits(&value) {
+                return Err(InterpreterError::FailureConstructingTupleWithType.into());
+            }
+            data_map.insert(name, value);
+        }
+        Self::new(expected.clone(), data_map)
+    }
+
+    pub fn get(&self, name: &str) -> Result<Value> {
+        self.data_map.get(name)
+            .cloned()
+            .ok_or_else(|| UncheckedError::NoSuchTupleField.into())
+    }
+
+    pub fn size(&self) -> Result<i128> {
+        self.type_signature.size()
+    }
+}
+
+impl fmt::Display for TupleData {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "(tuple")?;
+        for (name, value) in self.data_map.iter() {
+            write!(f, " ")?;
+            write!(f, "({} {})", &**name, value)?;
+        }
+        write!(f, ")")
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    #[test]
+    fn test_constructors() {
+        assert_eq!(
+            Value::list_with_type(
```

Comment:
Can we add a test with higher list dimensions?

Reply:
That following test (line 412) does use higher dimensions.